stencils: Add tests and documentation (#619)
JuliaParallel/Dagger.jl: docs/src/stencils.md

@assert collect(A) == expected_A
@assert collect(B) == expected_B_seq
```

## Operations on Multiple `DArray`s

You can read from and write to multiple `DArray`s within a single `@stencil` block, provided they have compatible chunk structures.

```julia
A = ones(Blocks(1, 1), Int, 2, 2)
B_multi = Dagger.fill(Blocks(1, 1), 2, Int, 2, 2) # Renamed to avoid conflict, corrected fill
C = zeros(Blocks(1, 1), Int, 2, 2)

Dagger.spawn_datadeps() do
    @stencil begin
        C[idx] = A[idx] + B_multi[idx] # Use the renamed B_multi
    end
end
@assert all(collect(C) .== 3)
```

## Example: Game of Life

The following demonstrates a more complex example: Conway's Game of Life.

```julia
# Ensure Plots and other necessary packages are available for the example
# using Plots

N = 27 # Size of one dimension of a tile
nt = 3 # Number of tiles in each dimension (results in nt x nt grid of tiles)
niters = 10 # Number of iterations for the animation

tiles = zeros(Blocks(N, N), Bool, N*nt, N*nt)
outputs = zeros(Blocks(N, N), Bool, N*nt, N*nt)

# Create a fun initial state (e.g., a glider and some random noise)
tiles[13, 14] = true
tiles[14, 14] = true
tiles[15, 14] = true

> tiles[15, 15] = true # Corrected glider part

Author: ```suggestion
tiles[15, 15] = true
```

tiles[14, 16] = true
# Add some random noise in one of the tiles

> # Make sure to use Dagger-compatible assignment if you were to modify chunks directly
> # For simplicity, direct array indexing is used here for initial setup.
> rand_tile_data = rand(Bool, N, N)
> # To assign this to a specific block, you'd typically work with chunks,
> # but for initial setup, direct indexing on the collected array or careful DArray construction is easier.
> # For this example, we'll simplify and assume direct modification is for setup.
> # A Dagger-idiomatic way for partial modification might involve map! or similar.
> # Here, we just modify the underlying array before it's heavily used by Dagger tasks if possible,
> # or use Dagger operations.
> # For collected view for setup:
> temp_tiles = collect(tiles) # This collect is fine for initial setup visualization/modification
> temp_tiles[(2N+1):3N, (2N+1):3N] .= rand_tile_data
> tiles = Dagger.distribute(temp_tiles, Blocks(N,N)) # Use distribute to create DArray from existing array

Author: ```suggestion
@view(tiles[(2N+1):3N, (2N+1):3N]) .= rand(Bool, N, N)
```

> # The animation part requires a graphical environment.
> # If running in a headless environment, you might comment out the @animate macro
> # and inspect `outputs` programmatically.
> # anim = @animate for _ in 1:niters
> #     Dagger.spawn_datadeps() do
> #         @stencil begin
> #             outputs[idx] = begin
> #                 nhood = @neighbors(tiles[idx], 1, Wrap())
> #                 live_neighbors = sum(nhood) - tiles[idx] # Subtract self if it was counted
> #                 if tiles[idx] # If current cell is alive
> #                     if live_neighbors < 2 || live_neighbors > 3
> #                         false # Dies by underpopulation or overpopulation
> #                     else
> #                         true  # Survives
> #                     end
> #                 else # If current cell is dead
> #                     if live_neighbors == 3
> #                         true  # Becomes alive by reproduction
> #                     else
> #                         false # Stays dead
> #                     end
> #                 end
> #             end
> #             tiles[idx] = outputs[idx] # Update tiles for the next iteration
> #         end
> #     end
> #     # heatmap(Int.(collect(outputs))) # Visualize (requires Plots.jl)
> # end
> # path = mp4(anim; fps=5, show_msg=true).filename # Save animation (requires Plots.jl)

Author: ```suggestion

anim = @animate for _ in 1:niters
    Dagger.spawn_datadeps() do
        @stencil begin
            outputs[idx] = begin
                nhood = @neighbors(tiles[idx], 1, Wrap())
                neighs = sum(nhood) - tiles[idx] # Sum neighborhood, but subtract own value
                if tiles[idx] && neighs < 2
                    0 # Dies of underpopulation
                elseif tiles[idx] && neighs > 3
                    0 # Dies of overpopulation
                elseif !tiles[idx] && neighs == 3
                    1 # Becomes alive by reproduction
                else
                    tiles[idx] # Keeps its prior value
                end
            end
            tiles[idx] = outputs[idx] # Update tiles for the next iteration
        end
    end
    heatmap(Int.(collect(outputs))) # Generate a heatmap visualization
end
path = mp4(anim; fps=5, show_msg=true).filename # Create an animation of the heatmaps over time
```
